### PR TITLE
NON-JIRA: Cleaning unused old datastore entities during retention pro…

### PR DESCRIPTION
**Pull request by radkomateusz**
…cess.

Process for removing entities about tables not existing for a long time and their backup entities.
After one month after removing last backup by retention process, the table entity and backup entities connected with them are deleted.

Hardcoded constants for managing retention are now moved to configuration.
Documentation and tests updated. Renaming filter name to not contain hardcoded value in class name.

### CODE REVIEW

File: tests/retention/test_table_retention.py [hunk 1]
```diff
@@ -164,3 +165,89 @@ def test_should_not_perform_retention_if_no_backups(self, delete_table):
...
+        # when
+        self.under_test.perform_retention(reference, table.key.urlsafe())
+
+        # then
+        self.assertTrue(table.key.get() is not None)
+        self.assertTrue(Backup.get_by_key(backup1.key) is not None)
+        self.assertTrue(Backup.get_by_key(backup2.key) is not None)
+
+    @freeze_time("2019-05-16")
+    def test_should_not_delete_table_and_backups_entities_if_table_not_exist_for_more_than_threshold_value_and_backups_are_not_deleted_yet(
```
Comment (marek-tabor) [inline tests/retention/test_table_retention.py]: missing test case when only 1 backup is deleted from the group

File: tests/backup/datastore/test_backup.py [hunk 2]
```diff
@@ -123,11 +123,45 @@ def test_that_get_all_backups_sorted_will_return_only_these_with_null_deleted_co
...
+            last_modified=datetime(2017, 02, 1, 16, 30),
+            created=datetime(2017, 02, 1, 16, 30),
+            dataset_id='backup_dataset',
+            table_id='backup1',
+            numBytes=1234,
+        )
+        backup1.put()
+        backup2 = Backup(
+            parent=table.key,
+            last_modified=datetime(2017, 02, 1, 16, 30),
```
Comment (marek-tabor) [inline tests/backup/datastore/test_backup.py]: 02 -> 2

File: tests/retention/test_table_retention.py (hunk 1 above)
Comment (radkomateusz) [reply tests/retention/test_table_retention.py]: there is no situation like removing one entity but leaving another. Everything stays, or everything is deleted (if all backups entities are marked as deleted and source table last seen is above grace deletion period)

File: tests/backup/datastore/test_backup.py (hunk 2 above)
Comment (radkomateusz) [reply tests/backup/datastore/test_backup.py]: done

File: src/retention/should_perform_datastore_cleanup_predicate.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+from src.commons.config.configuration import configuration
+
+
+class ShouldPerformDatastoreCleanupPredicate(object):
+
+    @staticmethod
+    def test(url_safe_key, existing_backups):
+        if existing_backups:
+            return False
+        if ShouldPerformDatastoreCleanupPredicate.\
```
Comment (marcin-kolda) [inline src/retention/should_perform_datastore_cleanup_predicate.py]: You can just return the value from method

File: src/retention/table_retention.py [hunk 4]
```diff
@@ -52,10 +52,19 @@ def __delete_backup_in_bq_and_update_datastore(self, backup):
...
+            error_message = "Could not delete backup '{}' error: {}: {}" \
                 .format(backup.table_id, type(ex), ex)
             logging.exception(error_message)
+
+    @staticmethod
+    def __delete_datastore_entities_for_given_table(url_safe_key, table_reference):
+        table_key = ndb.Key(urlsafe=url_safe_key)
+        backups_entities_to_delete = Backup.get_all_backups_sorted(table_key)
+        logging.info("Removing table and backup entites for table:'%s'", table_reference)
+        for backup in backups_entities_to_delete:
```
[the remark this answers is not in the document]
Comment (marcin-kolda) [reply src/retention/table_retention.py]: You could also try to delete all of them within transaction (but there might be some limits for number of entities)

File: tests/retention/test_table_retention.py [hunk 5]
```diff
@@ -164,3 +165,89 @@ def test_should_not_perform_retention_if_no_backups(self, delete_table):
 
         # then
         delete_table.assert_not_called()
+
+    @freeze_time("2019-05-16")
+    def test_should_delete_both_table_and_backups_entities_if_table_not_exist_for_a_long_time_and_any_not_deleted_backup_exists(
```
Comment (marcin-kolda) [inline tests/retention/test_table_retention.py]: test_should_delete_both_table_and_backups_entities_if_table_not_exist_for_a_long_time_and_any_not_deleted_backup_exists => test_should_delete_both_table_and_backups_entities_if_table_does_not_exist_for_a_long_time_and_there_is_no_not_deleted_backup

File: tests/retention/test_table_retention.py [hunk 6]
```diff
@@ -164,3 +165,89 @@ def test_should_not_perform_retention_if_no_backups(self, delete_table):
...
+        # when
+        self.under_test.perform_retention(reference, table.key.urlsafe())
+
+        # then
+        self.assertTrue(table.key.get() is None)
+        self.assertTrue(Backup.get_by_key(backup1.key) is None)
+        self.assertTrue(Backup.get_by_key(backup2.key) is None)
+
+    @freeze_time("2019-05-16")
+    def test_should_not_delete_table_and_backups_entities_if_table_not_exist_for_less_than_threshold_value_and_backups_are_already_deleted(
```
Comment (marcin-kolda) [inline tests/retention/test_table_retention.py]: test_should_not_delete_table_and_backups_entities_if_table_not_exist_for_less_than_threshold_value_and_backups_are_already_deleted => test_should_not_delete_table_and_backup_entities_if_table_does_not_exist_for_less_than_threshold_value_and_backups_are_already_deleted

**Comment by radkomateusz** [comment]
We decided to chose more safety option of optimisation, which doesn't require deleting anything.

**Pull request closed without merging**